// @mapbox/search-js-react/web/Search.tsx
import React, { useRef, useEffect, useState } from 'react';
import mapboxgl from 'mapbox-gl';

import { SearchBox } from '../src/components/SearchBox';
import { MapboxSearchBox } from '@mapbox/search-js-web';

const ACCESS_TOKEN = MAPBOX_ACCESS_TOKEN;

export function Search(): React.ReactElement {
  const [map, setMap] = useState(null);
  const [value, setValue] = useState('');

  const mapContainerRef = useRef(null);

  useEffect(() => {
    mapboxgl.accessToken = ACCESS_TOKEN;

    const map = new mapboxgl.Map({
      container: mapContainerRef.current,
      style: 'mapbox://styles/mapbox/streets-v11',
      center: [-73.943, 40.7789],
      zoom: 11
    });

    setMap(map);

    const mapSearchControl = new MapboxSearchBox();
    mapSearchControl.accessToken = ACCESS_TOKEN;
    map.addControl(mapSearchControl);

    // Clean up on unmount
    return () => map.remove();
  }, []);

  return (
    <div
      style={{
        height: '100%',
        width: '100%',
        display: 'flex',
        flexDirection: 'column'
      }}
    >
      <div style={{ margin: '24px 12px' }}>
        <SearchBox
          accessToken={ACCESS_TOKEN}
          value={value}
          onChange={(v) => {
            setValue(v);
            console.log('change', v);
          }}
          onSuggest={(res) => console.log('suggest', res)}
          onSuggestError={(err) => console.log('error', err)}
          onRetrieve={(res) => console.log('retrieve', res)}
          map={map}
        />
      </div>
      <br></br>
      <div
        className="map-container"
        style={{ height: '100%', width: '100%' }}
        ref={mapContainerRef}
      />
    </div>
  );
}
